apps/hash-frontend/src/pages/types/[[...type-kind]].page.tsx

```diff
@@ -0,0 +1,229 @@
+  "link-type",
+  "property-type",
+  "data-type",
+] as const;
+
+type ParsedQueryKindParam = (typeof parsedQueryParams)[number];
+
+type ParsedQueryParams = {
+  ["type-kind"]?: ParsedQueryKindParam[];
+};
+
+export const getStaticPaths: GetStaticPaths<ParsedQueryParams> = () => ({
+  paths: [
+    { params: { "type-kind": [] } },
+    ...parsedQueryParams.map((kind) => ({ params: { "type-kind": [kind] } })),
+  ],
+  fallback: false,
+});
+
+export type TabId = "all" | ParsedQueryKindParam;
+
+type TypesPageProps = {
+  currentTab: TabId;
+};
+
+export const getStaticProps: GetStaticProps<
+  TypesPageProps,
+  ParsedQueryParams
+> = ({ params }) => {
+  const currentTab = params?.["type-kind"]?.[0] ?? "all";
+
+  return { props: { currentTab } };
+};
+
+const contentMaxWidth = 1000;
+
+const TypesPage: NextPageWithLayout<TypesPageProps> = ({ currentTab }) => {
+  const latestEntityTypes = useLatestEntityTypesOptional({
+    includeArchived: true,
+  });
+
+  const latestNonLinkEntityTypes = useMemo(
+    () =>
+      latestEntityTypes?.filter(
+        (entityType) => !isLinkEntityType(entityType.schema),
+      ),
+    [latestEntityTypes],
+  );
+
+  const latestLinkEntityTypes = useMemo(
+    () =>
+      latestEntityTypes?.filter((entityType) =>
+        isLinkEntityType(entityType.schema),
+      ),
+    [latestEntityTypes],
+  );
+
+  const { queryPropertyTypes } = useBlockProtocolQueryPropertyTypes();
+  const { queryDataTypes } = useBlockProtocolQueryDataTypes();
+
+  const [latestPropertyTypes, setLatestPropertyTypes] =
+    useState<PropertyTypeWithMetadata[]>();
+
+  const [latestDataTypes, setLatestDataTypes] =
+    useState<DataTypeWithMetadata[]>();
+
+  const fetchPropertyTypes = useCallback(async () => {
+    await queryPropertyTypes({ data: {} }).then(
+      ({ data: propertyTypesSubgraph }) => {
+        if (propertyTypesSubgraph) {
+          setLatestPropertyTypes(
+            getRoots<PropertyTypeRootType>(propertyTypesSubgraph),
+          );
+        }
+      },
+    );
+  }, [queryPropertyTypes]);
+
+  const fetchDataTypes = useCallback(
+    async () =>
+      await queryDataTypes({ data: {} }).then(({ data: dataTypesSubgraph }) => {
+        if (dataTypesSubgraph) {
+          setLatestDataTypes(getRoots<DataTypeRootType>(dataTypesSubgraph));
+        }
+      }),
+    [queryDataTypes],
+  );
+
+  useEffect(() => {
+    void fetchPropertyTypes();
+    void fetchDataTypes();
+  }, [fetchPropertyTypes, fetchDataTypes]);
+
+  const allTypes = useMemo(
+    () =>
+      latestNonLinkEntityTypes &&
+      latestLinkEntityTypes &&
+      latestPropertyTypes &&
+      latestDataTypes
+        ? [
+            ...latestNonLinkEntityTypes,
+            ...latestLinkEntityTypes,
+            ...latestPropertyTypes,
+            ...latestDataTypes,
+          ]
+        : undefined,
+    [
+      latestNonLinkEntityTypes,
+      latestLinkEntityTypes,
+      latestPropertyTypes,
+      latestDataTypes,
+    ],
+  );
```

Comment:
Nit: this'd be a bit shorter – are there behaviour changes it causes that are undesirable?

Saves the checks for `allTypes` being defined further down also
```suggestion
  const allTypes = useMemo(
    () => [
      ...(latestNonLinkEntityTypes ?? []),
      ...(latestLinkEntityTypes ?? []),
      ...(latestPropertyTypes ?? []),
      ...(latestDataTypes ?? []),
    ],
    [
      latestNonLinkEntityTypes,
      latestLinkEntityTypes,
      latestPropertyTypes,
      latestDataTypes,
    ],
  );
```

Reply:
I had this previously, but I wanted `allTypes` to be `undefined` when it's items have not yet been fetched.

Reply:
Fair enough, thanks!